enhancements/installer/telemeter-metrics.md

```diff
@@ -0,0 +1,430 @@
+
+## Open Questions
+
+ > 1. How to filter to collect only real user usage and prevent OpenShift developer/Red Hat internal users from skewing metrics? Perhaps filter by version?
+ > 2. Is it possible to determine whether a user modified an ignition config or manifest asset by comparing the asset the installer would generate to the one found on disk?
+ > 3. How to authenticate using pull secret.
+ 
+## Summary
+Telemetry collects metrics from running clusters, but no metrics are obtained from the installer binary.
+This enhancement proposes that the installer would push metrics related to user workflows, installation statistics, and feature usage. The installer would be instrumented to collect Prometheus label key-value pairs, convert them to metric samples and push those metrics to an [aggregation pushgateway][weaveworks-aggregation-pushgateway].
+This enhancement focuses on the design of the metrics to be pushed to Prometheus.
+
+## Motivation
+
+Obtaining metrics from the installer would allow engineers and product managers to understand different user workflows, environments, experiences, and difficulties, as well as an indication of feature usage.  By pushing these metrics to the Telemeter, OpenShift members would be able to create a variety of queries related to these aspects.
+
+### Goals
+
+1. Design meaningful metrics to collect user experience and behavior with the installer
+
+### Non-Goals
+
+1. Collected metrics will not allow analysis of individual installer invocations and will not display detailed data such as log or error messages. We won't be able to answer "how long did an individual installation take?" but rather 95% of installs finish within, say, 40 minutes.
+2. Installer instrumentation/internal implementation will be discussed in a separate design/enhancement
+3. Pushgateway implementation will be discussed in a separate design/enhancement
+
+## Proposal
+
+### User Stories
+
+#### Telemetry Analyst
+Some examples:
+
+As a telemetry analyst I would like to create queries which analyze aggregate installation data to determine how long the installation process takes in total and broken down by different stages so that I can better understand user experience, expose any abnormalities, and compare different versions of the installer.
+
+As a telemetry analyst I would like to create queries which show aggregate counts of failure by stage of installation so that we can better understand where users are experiencing problems. (This will not provide  specific error messages/logs).
+
+As a telemetry analyst I would like to see how often users modify ignition configs or manifests so that I can better understand how users are interacting with the installer. [(see open question #3)](open-questions)
+
+As a telemetry analyst I would like to see how often the installer is being run on certain operating systems so I can better understand our user environments and needs.
```

Comment:
Per our convo over in Jira, would love to see OPENSHIFT_INSTALL_INVOKER reported so we can see hive provisions in the wild.

Reply:
@dgoodwin the value for INVOKER is `user` if unset or the value from the env var. How can we bound the cardinality but still provide value in the metrics? Would it be enough to do a binary classification to say "these are all the installs where INVOKER is set (not `user`)"? 

If we need more detail than that, can you provide some input on how the `INVOKER` value is formatted?

Reply:
This doesn't strike me as a use case where cardinality would be an issue. We should only be dealing with a handful of values, user, hive, managed azure... maybe a dozen? It's not something that would be driven by user input or random values in any normal operation. Is there a concern over some malicious or overly enthusiastic user finding the var and setting it to random values?

Just to clarify are we expecting a metric like openshift_install_invoker{type=user} = 1? 

Could we trim this down to type=unknown if we don't recognize the type in the aggregation.

Reply:
>  maybe a dozen? It's not something that would be driven by user input or random values in any normal operation.

sounds good. I am (or was, until now) ignorant of the values. Are these values defined in hive code somewhere? I think the best way it would work is if hive & the installer had some sort of shared enumeration of potential values, similar in concept to how values (for e.g. operating system) are enumerated here: https://github.com/openshift/enhancements/blob/2196381b04834a96b15a739611be14d46a200d9b/enhancements/installer/telemeter-metrics.md#label-values

> . Is there a concern over some malicious or overly enthusiastic user finding the var and setting it to random values?

It seems like an unlikely threat but is so easy to guard against we should definitely do it to avoid any repercussions to production systems. If it is a nuisance to have a shared enumeration between installer & hive we could rethink it.

> Just to clarify are we expecting a metric like openshift_install_invoker{type=user} = 1?

No. I was thinking more along the lines of we would add it to any or all of the metrics suggested here that would be useful. If we added invoker to the cluster creation metric we could have samples like this for for the 15 minute or less installation bucket:
```
cluster_installation_create_bucket{invoker="user",platform="azure",result="success",version="4.6",os="linux",le="15"} 0
cluster_installation_create_bucket{invoker="managed-azure",platform="azure",result="success",version="4.6",os="linux",le="15"} 3
```
This would allow us to use all these metrics and select along an invoker dimension.

Reply:
> This would allow us to use all these metrics and select along an invoker dimension.

No need for this.  A single metric with the invoker label can be joined with other metrics at query-time using `_id` (the cluster's Telemetry ID).

Reply:
I would expect Hive would just have one hardcoded value we submit as invoker for any cluster install started via Hive. I'd really just be looking for trends, how many installs are being done through Hive's over time and their success rates.

Reply:
we also using the OPENSHIFT_INSTALL_INVOKER='assisted-installer' for clusters that will be created in such manner